agent/tcs/client/client.go

```diff
@@ -0,0 +1,138 @@
+// Copyright 2014-2015 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//	http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package tcsclient
+
+import (
+	"bytes"
+	"errors"
+	"net/http"
+	"time"
+
+	"github.com/aws/amazon-ecs-agent/agent/ecs_client/authv4"
+	"github.com/aws/amazon-ecs-agent/agent/ecs_client/authv4/credentials"
+	"github.com/aws/amazon-ecs-agent/agent/logger"
+	"github.com/aws/amazon-ecs-agent/agent/stats"
+	"github.com/aws/amazon-ecs-agent/agent/tcs/model/ecstcs"
+	"github.com/aws/amazon-ecs-agent/agent/wsclient"
+	"github.com/gorilla/websocket"
+)
+
+var log = logger.ForModule("tcs client")
+
+// clientServer implements wsclient.ClientServer interface for metrics backend.
+type clientServer struct {
+	statsEngine  stats.Engine
+	publishTimer *timer
+	wsclient.ClientServerImpl
+	signer authv4.HttpSigner
+}
+
+// New returns a client/server to bidirectionally communicate with the backend.
+// The returned struct should have both 'Connect' and 'Serve' called upon it
+// before being used.
+func New(url string, region string, credentialProvider credentials.AWSCredentialProvider, acceptInvalidCert bool, statsEngine stats.Engine, publishMetricsInterval time.Duration) wsclient.ClientServer {
+	cs := &clientServer{
+		statsEngine:  statsEngine,
+		publishTimer: newTimer(publishMetricsInterval, publishMetrics),
+		signer:       authv4.NewHttpSigner(region, wsclient.ServiceName, credentialProvider, nil),
+	}
+	cs.URL = url
+	cs.Region = region
+	cs.CredentialProvider = credentialProvider
+	cs.AcceptInvalidCert = acceptInvalidCert
+	cs.ServiceError = &tcsError{}
+	cs.RequestHandlers = make(map[string]wsclient.RequestHandler)
+	cs.TypeDecoder = &TcsDecoder{}
+	return cs
+}
+
+// Serve begins serving requests using previously registered handlers (see
+// AddRequestHandler). All request handlers should be added prior to making this
+// call as unhandled requests will be discarded.
+func (cs *clientServer) Serve() error {
+	log.Debug("Starting websocket poll loop")
+	if cs.Conn == nil {
+		return errors.New("nil connection")
+	}
+
+	// Start the timer function to publish metrics to the backend.
+	if cs.statsEngine != nil {
+		cs.publishTimer.start(cs)
```

Comment:
Might make sense to instantiate a 'time.NewTicker' here, toss publishmetrics on it in a goroutine, and stop it in Close as you do now. That would let you get rid of your timer type and would still behave equivalently I think.